### PR TITLE
Code Editor: Add documentation tooltips

### LINKED ISSUE #1393: Implement script editor description hint on hover a symbol/word
**Describe the project you are working on:**
Description hint on hover for the text edit(https://github.com/ThakeeNathees/godot/tree/TextEdit-hover-hint)

**Describe the problem or limitation you are having in your project:**
it's one of the essential feature for a fully featured script editor and currently we lack of. Also the implementation of GDScript doc comments (https://github.com/godotengine/godot/pull/41095) it'll improve the user experience.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
when a symbol is hovered by the mouse, the text editor emits a signal to validate the symbol (like when it's being hovered with a ctrl), and when validating for that symbol it'll add some additional data (like description, type, class_name, return_type, arguments, etc.,) and display them at the text edit's draw call.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
A working example (this will work with doc comments when it'll be merged)
hint2

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
No, It would be a great usability improvement and can't be worked around with the script.

**Is there a reason why this should be core and not an add-on in the asset library?:**
there is no way to implement this with an add-on

*Bugsquad edit (keywords for easier searching): tooltip*

### PR DESCRIPTION
**Pull request by dalexeev**
* Closes godotengine/godot-proposals#1393.
* Supersedes #63908.
* Supersedes #80044.

<details>
<summary>Screenshots</summary>

</details>

### CODE REVIEW

**Comment by KoBeWi** [comment]
#80044 also implements tooltips for identifiers, so it's not fully superseded. It would be nice if hovering a variable would show its type at least.

From quick testing:
- The cursor changes shape as soon as tooltip appears and clicking outside the tooltip only closes it, instead of moving caret or selecting.
- When you scroll, the tooltip moves with the view.

**Comment by Atenvardo** [comment]
Amazing feature, hoping we get this built in soon. Great work!

**Comment by KoBeWi** [comment]
Tested again and it works great now. One major issue is that when moving cursor *very slightly* the tooltip will disappear, even if you move it over the same symbol. It will open again, but the frequent blinking is bothersome.

Also if you hover a symbol and then move the cursor outside script editor, the tooltip will still appear.

https://github.com/user-attachments/assets/b4c725c1-2f5e-45cc-942e-4749b1c36ced

**Comment by dalexeev** [comment]
Ready for testing/review.

* Depends on #95044.
* This would go well with #95821.
* It also makes sense to separate <kbd>Ctrl+LMB</kbd> functions. Say <kbd>Ctrl+LMB</kbd> - open the docs and <kbd>Alt+LMB</kbd> - open the code.

Along the way I noticed/fixed a few bugs in `gdscript_editor.cpp`, `gdscript_docgen.cpp`, and `script_text_editor.cpp`, but there are still some things that make sense to fix separately, as I'm afraid that this might be a significant refactoring (enums, FQCN).

> Also if you hover a symbol and then move the cursor far away, even outside script editor, the tooltip will still appear.

Fixed, thanks.

> One major issue is that when moving cursor _very slightly_ the tooltip will disappear, even if you move it over the same symbol. It will open again, but the frequent blinking is bothersome.

I understand that Visual Studio Code does it better (the tooltip doesn't disappear until the mouse leaves the symbol), but I'd rather leave it as is for now. This is consistent with the behavior in the inspector and signal dock, and it's also the same as in Kate.



But most importantly, it's technically hard to do. The tooltip checks if the mouse has left the target. This works for simple controls, but not for `Tree` and `CodeEdit`, which consist of many visual parts that are not nodes. Also, `TextEdit`/`CodeEdit` allows you to get the symbol at a given position as a string, but has no API to get positional information so we can determine whether it was the same symbol or another such identifier on the same line.

I'd suggest leaving non-critical issues for later PRs, as I don't want to further affect unrelated parts of the engine. In my opinion, this is not a very critical issue, I tried to make it so that the tooltip does not interfere with keys, clicks, scrolling and text selection.

Another usability issue is that `TextEdit` finds a symbol outside the line:


But the same happens if you hold down <kbd>Ctrl</kbd>:

**Comment by KoBeWi** [comment]
>This is consistent with the behavior in the inspector and signal dock, and it's also the same as in Kate.

For some reason I find it more difficult to see tooltips in code editor than in the inspector. Maybe the minimum movement is too small?

**Comment by dalexeev** [comment]
> Hovering over variable definition does not show a tooltip. It could be useful when using inference, so e.g. you can hover `var something := method()` and know the type of `something` at the source, without having to find its usage.

Looks like it's not implemented in the parser:

https://github.com/godotengine/godot/blob/19e003bc08ac2fb14fc2434bde9530387afd994c/modules/gdscript/gdscript_parser.h#L1294-L1298

When I hover over the variable declaration identifier I get `{type = GDScriptParser::COMPLETION_NONE, current_class = 0x0, current_function = 0x0, current_suite = 0x0, current_line = -1, current_argument = -1, builtin_type = Variant::VARIANT_MAX, node = 0x0, base = 0x0, parser = 0x0}`.

This and some other issues with incorrect hints are material for future PRs.

> For some reason I find it more difficult to see tooltips in code editor than in the inspector. Maybe the minimum movement is too small?

I can't reproduce it on Linux. I encountered this during development, but at some point the bug disappeared.

File: scene/gui/code_edit.cpp [hunk 1]
```diff
@@ -3753,6 +3789,13 @@ CodeEdit::CodeEdit() {
 	set_gutter_custom_draw(gutter_idx, callable_mp(this, &CodeEdit::_fold_gutter_draw_callback));
 	gutter_idx++;
 
+	/* Symbol tooltip */
+	symbol_tooltip_timer = memnew(Timer);
+	symbol_tooltip_timer->set_wait_time(GLOBAL_GET("gui/timers/tooltip_delay_sec"));
```
Comment (KoBeWi) [inline scene/gui/code_edit.cpp:3794]: Would be nice to update it with `settings_changed` signal.

File: editor/editor_help.cpp [hunk 2]
```diff
@@ -99,6 +100,10 @@ const Vector<String> classes_with_csharp_differences = {
 };
 #endif
 
+static const String nbsp = String::chr(160);
+static const String nbsp_equal_nbsp = nbsp + "=" + nbsp;
+static const String colon_nbsp = ":" + nbsp;
```
Comment (KoBeWi) [inline editor/editor_help.cpp:105]:
```suggestion
static constexpr String nbsp = String::chr(160);
static constexpr String nbsp_equal_nbsp = nbsp + "=" + nbsp;
static constexpr String colon_nbsp = ":" + nbsp;
```
Preferred for global constants.

File: editor/editor_help.cpp [hunk 3]
```diff
@@ -166,6 +188,39 @@ static String _contextualize_class_specifier(const String &p_class_specifier, co
 	return p_class_specifier.substr(p_edited_class.length() + 1);
 }
 
+/// EditorHelpBit ///
+
+// TODO: this is sometimes used directly as doc->something, other times as EditorHelp::get_doc_data(), which is thread-safe.
```
Comment (KoBeWi) [inline editor/editor_help.cpp:193]:
```suggestion
// TODO: This is sometimes used directly as doc->something, other times as EditorHelp::get_doc_data(), which is thread-safe.
```

**Comment by KoBeWi** [comment]
Looks like the description can't be determined in complex code:
image
Even though Ctrl+Click jumps to the member and it has a description.

I assume this is related to #95821?

**Review by KoBeWi — APPROVED** [review]
The base functionality works fine. I agree it can be improved later if needed.
I checked the code of editor and gui in the second commit and also looks ok.

File: editor/editor_help.cpp (hunk 2 above)
Comment (dalexeev) [reply editor/editor_help.cpp:105]: `String` has no `constexpr` constructors.

File: editor/editor_help.cpp [hunk 4]
```diff
@@ -3691,51 +3952,107 @@ void EditorHelpBit::_notification(int p_what) {
 }
 
 void EditorHelpBit::parse_symbol(const String &p_symbol) {
-	const PackedStringArray slices = p_symbol.split("|", true, 2);
-	ERR_FAIL_COND_MSG(slices.size() < 3, "Invalid doc id. The expected format is 'item_type|class_name|item_name'.");
+	const PackedStringArray slices = p_symbol.split("|", true, 3);
+	ERR_FAIL_COND_MSG(slices.size() < 3, "Invalid doc id: The expected format is \"item_type|class_name|item_name[|item_data]\".");
```
Comment (AThousandShips) [inline editor/editor_help.cpp:3956]:
```suggestion
	ERR_FAIL_COND_MSG(slices.size() < 3, R"(Invalid doc id: The expected format is "item_type|class_name|item_name[|item_data]".)");
```

**Review by AThousandShips — APPROVED** [review]
LGTM

**Review by Calinou — APPROVED** [review]
Tested locally, it works as expected.

Note that you can select and drag code that's behind the tooltip. You also can't scroll the scrollbar that is present for long descriptions, but as the above comments say, it may be better to leave it as-is for now.

Regarding the spawn position of the tooltip, it should probably spawn *above* the hovered line instead of below. It should also be done at a fixed position (corner aligned to the beginning of the hovered word), not taking the mouse cursor's position into account. This way, you can still look at the lines following the hovered line, which is usually more relevant in most situations. Of course, if there isn't enough space to show the tooltip above the hovered line, then it should be shown below as is done currently.

See VS Code for an example:

image

Also, there could be a keyboard shortcut to show the tooltip the cursor is under, so that you don't have to open the full documentation page if you're just interested in a single method's description. VS Code uses a dual combination for this (<kbd>Ctrl + K</kbd> then <kbd>Ctrl + I</kbd>) but I'd prefer something less cumbersome.

**Comment by dalexeev** [comment]
> You also can't scroll the scrollbar that is present for long descriptions

Are you using single window mode? This bug exists from the start. Scrolling should work in multi-window mode.

> Regarding the spawn position of the tooltip, it should probably spawn _above_ the hovered line instead of below.

I think it's easy to fix, just change the priority in `popup_under_cursor()`. I'll take a look later.

> It should also be done at a fixed position (corner aligned to the beginning of the hovered word), not taking the mouse cursor's position into account.

This is more difficult to do, I wrote above why. I would prefer to leave it for the next iterations.

> Also, there could be a keyboard shortcut to show the tooltip the cursor is under, so that you don't have to open the full documentation page if you're just interested in a single method's description. VS Code uses a dual combination for this (Ctrl + K then Ctrl + I) but I'd prefer something less cumbersome.

I'm not sure I understood what you meant, but we could add a button to the tooltip title or make the symbol name a link:

**Comment by dalexeev** [comment]
> > Regarding the spawn position of the tooltip, it should probably spawn _above_ the hovered line instead of below.
> 
> I think it's easy to fix, just change the priority in `popup_under_cursor()`. I'll take a look later.

Done, but I really don't like how it looks. If for short descriptions it looks more or less ok, then for long descriptions it is not convenient. All symbol information as well as brief description (for classes) are located too far from the cursor.

<details>
<summary>Screenshots</summary>




</details>

Maybe I am too used to tooltips in Kate, but it seems to me that VSC-way is not suitable here. And I don't really understand what problem this solves? Don't overlap the code below? But the tooltip disappears when you start typing and it is impossible to say which context will be more useful for you at the moment, with the same success it could be the code above.

**Comment by Mickeon** [comment]
I think of my own personal bias that the tooltip's background is not distinct at all from the Script editor's, at least by default. That, combined with not much inner padding, makes the tooltip look "noiser" than it should, I feel.

**Comment by Repiteo** [comment]
Thanks!

**Comment by Calinou** [comment]
> I'm not sure I understood what you meant, but we could add a button to the tooltip title or make the symbol name a link:

This is what I mean:

https://github.com/user-attachments/assets/a34c518f-db95-434d-83f7-9a68ff79df5d

At the end of the video, I press <kbd>Ctrl + K</kbd> then <kbd>Ctrl + I</kbd> to make the tooltip appear without having to touch the mouse.

**Comment by JekSun97** [comment]
Perhaps in the future it will be possible to add this to the help on scripting, for links on functions.

**Comment by filkata** [comment]
Is a side effect of this that ctrl+click on a function call in my own GDscript code always leads to the documenation rather than the code for the function itself like before? I would like to know if there is a way to get the old behaviour as this severely afects my workflow.

**Comment by KoBeWi** [comment]
#100707

**Comment by GustJc** [comment]
Love this feature! It really improves the gdscript experience.

But I'm guessing there is no way to enable the tooltips to show up for the auto-complete options yet. 
That would be all that is missing now to make it really great.

Is that feature planned or tracked anywhere?
I think I found it here https://github.com/godotengine/godot-proposals/issues/408

So its okay to assume it won't go to 4.4 at this time? (as I think the beta release means its entering feature freeze?)

https://github.com/user-attachments/assets/96aaf67a-3a7f-4b2d-86fe-65a0974c8e28

**Comment by KoBeWi** [comment]
> So its okay to assume it won't go to 4.4 at this time? (as I think the beta release means its entering feature freeze?)

Yes, the beta arrives very soon and no one even opened a PR for that.

**Comment by dalexeev** [comment]
* @GustJc See #101370.

**Comment by quitelight** [comment]
This feature is buggy as hell.

1) Provide option to disable the tooltips in editor options.
2) Tooltips is very buggy on MACOS, it disappears upon hover, cannot scroll and cannot select tooltip text or click on links.

**Comment by dalexeev** [comment]
> 1. Provide option to disable the tooltips in editor options.

* Thanks for the suggestion, I implemented the editor setting and opened a PR, see #102608.

> IMO don't release a broken feature that doesn't work in unison across all platforms.

Many Windows and Linux users seem to be happy with the feature, in my opinion it would be unfair to drop this QoL change for all users because of MacOS issues. Unfortunately, we don't have many MacOS contributors, so support for this platform is worse in some cases.

The implementation does not affect platform-dependent code much, it just exposed existing issues for some platforms and single-window mode support. Which can also appear in user projects.

Most contributors are volunteers, many features are supported on a best-effort basis. So dropping this feature for 4.4 would not guarantee that the MacOS issues would be fixed for 4.5, most likely everything would remain as is.

As long as this feature does not cause major issues (data loss, inability to use other features, etc.), it is not a release blocker. So let's release the feature with its minor issues and hope that someone in the future will notice and fix them.

**Comment by akien-mga** [comment]
> Tooltips is very buggy on MACOS, it disappears upon hover, cannot scroll and cannot select tooltip text or click on links.

Could you open bug report(s) for this? We're currently not aware of such issues on macOS, so indeed if they're a bother and need to be dealt with before a stable release, we need users to report them so we can fix the bugs.

**Comment by quitelight** [comment]
@dalexeev  Yeah totally understand, and definitely appreciate all the effort!

**Comment by eimfach** [comment]
The Changelogs states in the section of the new tooltips: "This includes any notes you have written yourself using the new documentation system." - what is meant with "new documentation system" ? What is it ?

**Comment by KoBeWi** [comment]
Documentation tooltips for GDScript. It's actually not very new anymore.

**Pull request merged**